File: Ruby/leetcode/shift-2d-grid/solution.rb
# @param {Integer[][]} grid
# @param {Integer} k
# @return {Integer[][]}
def shift_grid(grid, k)
  m = grid.size
  n = grid[0].size
  total = m * n
  k %= total
  return grid if k == 0

  flat = grid.flatten
  shifted = flat[-k..-1] + flat[0...-k]

  result = Array.new(m) { Array.new(n) }
  (0...total).each do |idx|
    row = idx / n
    col = idx % n
    result[row][col] = shifted[idx]
  end

  result
end

File: CSharp/leetcode/shift-2d-grid/solution.cs
public class Solution {
    public IList<IList<int>> ShiftGrid(int[][] grid, int k) {
        int m = grid.Length;
        int n = m > 0 ? grid[0].Length : 0;
        int total = m * n;
        var result = new List<IList<int>>();
        for (int i = 0; i < m; i++) result.Add(new List<int>(new int[n]));
        if (total == 0) return result;
        k %= total;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int pos = i * n + j;
                int np = (pos + k) % total;
                result[np / n][np % n] = grid[i][j];
            }
        }
        return result;
    }
}
